File: JAVA/Word Break.java
import java.util.*;
public class Solution {
    private int f[][] = null;
    public boolean wordBreak(String s, Set<String> dict) {
        int len = s.length();
        f = new int[len][len]; // 0 for unvisited, -1 for false, 1 for true
        return wordBreak(s, dict, 0, len-1);
    }
     
    private boolean wordBreak(String s, Set<String> dict, int i, int j) {
        if (f[i][j] == 1) return true;
        if (f[i][j] == -1) return false;
        String s0 = s.substring(i, j + 1);
        if (dict.contains(s0)) {
            f[i][j] = 1;
            return true;
        }
        for (int k = i + 1; k <= j; k++) {
            if (wordBreak(s, dict, i, k-1) && wordBreak(s, dict, k, j)) {
                f[i][j] = 1;
                return true;
            }
        }
        f[i][j] = -1;
        return false;
    }
}

File: Word Break.cpp
class Solution {
public:
    bool wordBreak(string s, unordered_set<string> &dict) {
        vector<vector<int> > canBreak; // 0 for unvisited, 1 for true, -1 for false
        int N = s.size();
        canBreak.resize(N);
        for (int i = 0; i < N; i++)
        {
            canBreak[i].resize(N);
        }
        return wordBreakRe(s, dict, canBreak, 0, N - 1);
    }
     
    bool wordBreakRe(string &s, unordered_set<string> &dict, vector<vector<int> > &canBreak, int start, int end)
    {
        if (canBreak[start][end] != 0)
            return (canBreak[start][end] == 1 ? true : false);
        string sub = s.substr(start, end - start + 1);
        if (dict.find(sub) != dict.end())
        {
            canBreak[start][end] = 1;
            return true;
        }
        for (int i = start; i < end; i++)
        {
            if (wordBreakRe(s, dict, canBreak, start, i) && 
                wordBreakRe(s, dict, canBreak, i + 1, end))
            {
                canBreak[start][end] = 1;
                return true;
            }
        }
        canBreak[start][end] = -1;
        return false;
    }
};
